core/src/SYCL/Kokkos_SYCL_Instance.cpp

```diff
@@ -302,6 +308,12 @@ template void SYCLInternal::fence_helper<sycl::event>(sycl::event&,
                                                       const std::string&,
                                                       uint32_t);
 
+// This function cycles through a pool of USM allocations for functors
+SYCLInternal::IndirectKernelMem& SYCLInternal::get_indirect_kernel_mem() {
+  m_pool_next = (m_pool_next + 1) % m_usm_pool_size;
+  return m_indirectKernelMem[m_pool_next];
+}
```

Comment:
We likely want this to be thread-safe and update m_pool_next atomically.

Reply:
Good point - done.